sprocs/sync_assessments.sql

```diff
@@ -0,0 +1,303 @@
+                    INSERT INTO assessment_questions AS aq (
+                        number,
+                        max_points,
+                        init_points,
+                        points_list,
+                        force_max_points,
+                        tries_per_variant,
+                        deleted_at,
+                        assessment_id,
+                        question_id,
+                        alternative_group_id,
+                        number_in_alternative_group
+                    ) VALUES (
+                        (assessment_question->>'number')::integer,
+                        (assessment_question->>'max_points')::double precision,
+                        (assessment_question->>'init_points')::double precision,
+                        (SELECT ARRAY_AGG(points)::double precision[] FROM JSONB_ARRAY_ELEMENTS_TEXT(COALESCE(assessment_question->>'points_list', '[]')::jsonb) points)::double precision[],
+                        (assessment_question->>'force_max_points')::boolean,
+                        (assessment_question->>'tries_per_variant')::integer,
+                        NULL,
+                        new_assessment_id,
+                        (assessment_question->>'question_id')::bigint,
+                        new_alternative_group_id,
+                        (assessment_question->>'number_in_alternative_group')::integer
+                    ) ON CONFLICT (question_id, assessment_id) DO UPDATE
+                    SET
+                        number = EXCLUDED.number,
+                        max_points = EXCLUDED.max_points,
+                        points_list = EXCLUDED.points_list,
+                        init_points = EXCLUDED.init_points,
+                        force_max_points = EXCLUDED.force_max_points,
+                        tries_per_variant = EXCLUDED.tries_per_variant,
+                        deleted_at = EXCLUDED.deleted_at,
+                        alternative_group_id = EXCLUDED.alternative_group_id,
+                        number_in_alternative_group = EXCLUDED.number_in_alternative_group,
+                        question_id = EXCLUDED.question_id
+                    RETURNING aq.id INTO new_assessment_question_id;
+                    new_assessment_question_ids := array_append(new_assessment_question_ids, new_assessment_question_id);
+                END LOOP;
+            END LOOP;
+            zone_index := zone_index + 1;
+        END LOOP;
+
+        -- Delete excess zones for this assessment
+        DELETE FROM zones
+        WHERE
+            assessment_id = new_assessment_id
+            AND number > jsonb_array_length(assessment->'zones');
+
+        -- Delete excess alternative groups for this assessment
+        DELETE FROM alternative_groups
+        WHERE
+            assessment_id = new_assessment_id
+            AND ((number < 1) OR (number > (assessment->>'lastAlternativeGroupNumber')::integer));
```

Comment:
The tricky thing here is that the alternative groups are in an array of arrays so we can't just use the length of something. We'll need to do some kind of calculation somewhere as we loop over the outer array. I guess I don't have a strong preference if we do this in JS and forward it into SQL or if it's done in the SQL. In this case, since I was computing the numbers in JS, I computed the last number there too.

Reply:
Ah, excellent point, I had missed this. All good!